app/models/match.rb

```diff
@@ -97,6 +97,12 @@ def has_points?
     points_home.present? && points_away.present?
   end
 
+  def opponent_of(participant)
+    return team_home_recursive if participant == team_away
```

Comment:
shouldn't you test for `team_home_recursive if participant == team_away_recursive`?